File: js/src/transport.js
import fetch from 'node-fetch';

export const RECEIPT_HEADERS = [
  'cache-control',
  'content-encoding',
  'content-length',
  'content-type',
  'etag',
  'last-modified',
  'location',
];

function selectedHeaders(headers) {
  const selected = {};
  for (const name of RECEIPT_HEADERS) {
    const value =
      typeof headers?.get === 'function' ? headers.get(name) : headers?.[name];
    if (value !== null && value !== undefined) {
      selected[name] = String(value);
    }
  }
  return selected;
}

function classifyTransportError(error, signal) {
  if (signal?.aborted || error?.name === 'AbortError') {
    return 'cancelled';
  }
  if (error?.name === 'TimeoutError' || error?.type === 'request-timeout') {
    return 'timeout';
  }
  if (error instanceof TypeError) {
    return 'cors_or_transport';
  }
  return 'transport';
}

export class CaptureTransportError extends Error {
  constructor(error, url, signal) {
    super(error?.message || String(error));
    this.name = 'CaptureTransportError';
    this.cause = error;
    this.diagnostics = {
      outcome: 'error',
      errorKind: classifyTransportError(error, signal),
      sourceUrl: url,
      error: this.message,
    };
  }
}

/** Default caller-replaceable transport. Dropping/aborting `signal` cancels fetch. */
export function fetchTransport({ url, method = 'GET', headers, signal }) {
  return fetch(url, { method, headers, signal });
}

/**
 * Capture an HTTP response without decoding its body.
 *
 * A custom transport receives `{url, method, headers, signal}` and may return
 * either a Fetch `Response` or an already materialized receipt. The returned
 * body is a Buffer containing the exact response bytes.
 */
export async function captureResponse(
  url,
  { transport = fetchTransport, signal, headers = {} } = {}
) {
  if (!url) {
    throw new Error('Missing URL parameter');
  }
  const request = { url, method: 'GET', headers, signal };
  try {
    const response = await transport(request);
    if (
      response?.body !== null &&
      response?.body !== undefined &&
      response?.finalUrl &&
      response?.diagnostics
    ) {
      return { ...response, body: Buffer.from(response.body) };
    }
    if (
      !response ||
      (typeof response.arrayBuffer !== 'function' &&
        typeof response.text !== 'function')
    ) {
      throw new TypeError(
        'Transport must return a Response or response receipt'
      );
    }
    const body =
      typeof response.arrayBuffer === 'function'
        ? Buffer.from(await response.arrayBuffer())
        : Buffer.from(await response.text());
    return {
      body,
      finalUrl: response.url || url,
      status: response.status,
      headers: selectedHeaders(response.headers),
      diagnostics: { outcome: 'response' },
    };
  } catch (error) {
    if (error instanceof CaptureTransportError) {
      throw error;
    }
    throw new CaptureTransportError(error, url, signal);
  }
}

File: rust/src/transport.rs
//! Caller-owned HTTP transport and exact-byte response receipts.

use std::{collections::BTreeMap, future::Future, pin::Pin};

use serde::{Deserialize, Serialize};

/// Headers retained in portable response receipts.
pub const RECEIPT_HEADERS: [&str; 7] = [
    "cache-control",
    "content-encoding",
    "content-length",
    "content-type",
    "etag",
    "last-modified",
    "location",
];

/// Owned request passed to an injectable transport.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TransportRequest {
    pub url: String,
    pub method: String,
    pub headers: BTreeMap<String, String>,
}

/// Structured transport outcome attached to a successful receipt.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TransportDiagnostics {
    pub outcome: String,
}

impl TransportDiagnostics {
    #[must_use]
    pub fn response() -> Self {
        Self {
            outcome: "response".to_string(),
        }
    }
}

/// Undecoded response bytes and provenance metadata.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ResponseReceipt {
    pub body: Vec<u8>,
    pub final_url: String,
    pub status: u16,
    pub headers: BTreeMap<String, String>,
    pub diagnostics: TransportDiagnostics,
}

/// Structured failure when no HTTP response receipt exists.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, thiserror::Error)]
#[error("{message}")]
#[serde(rename_all = "camelCase")]
pub struct TransportError {
    pub kind: String,
    pub message: String,
    pub source_url: String,
}

pub type TransportFuture<'a> =
    Pin<Box<dyn Future<Output = std::result::Result<ResponseReceipt, TransportError>> + Send + 'a>>;

/// Injectable asynchronous transport.
///
/// Cancellation is explicit through normal Rust future lifetime semantics:
/// dropping the future returned by [`Transport::execute`] cancels the request.
pub trait Transport: Send + Sync {
    fn execute(&self, request: TransportRequest) -> TransportFuture<'_>;
}

impl<F, Fut> Transport for F
where
    F: Fn(TransportRequest) -> Fut + Send + Sync,
    Fut: Future<Output = std::result::Result<ResponseReceipt, TransportError>> + Send + 'static,
{
    fn execute(&self, request: TransportRequest) -> TransportFuture<'_> {
        Box::pin(self(request))
    }
}

/// Default reqwest implementation. Callers may supply an already configured client.
#[cfg(feature = "runtime")]
#[derive(Debug, Clone)]
pub struct ReqwestTransport {
    client: reqwest::Client,
}

#[cfg(feature = "runtime")]
impl ReqwestTransport {
    #[must_use]
    pub const fn new(client: reqwest::Client) -> Self {
        Self { client }
    }
}

#[cfg(feature = "runtime")]
impl Default for ReqwestTransport {
    fn default() -> Self {
        Self::new(reqwest::Client::new())
    }
}

#[cfg(feature = "runtime")]
impl Transport for ReqwestTransport {
    fn execute(&self, request: TransportRequest) -> TransportFuture<'_> {
        Box::pin(async move {
            let method =
                reqwest::Method::from_bytes(request.method.as_bytes()).map_err(|error| {
                    TransportError {
                        kind: "invalid_request".to_string(),
                        message: error.to_string(),
                        source_url: request.url.clone(),
                    }
                })?;
            let mut builder = self.client.request(method, &request.url);
            for (name, value) in &request.headers {
                builder = builder.header(name, value);
            }
            let response = builder.send().await.map_err(|error| TransportError {
                kind: if error.is_timeout() {
                    "timeout"
                } else if error.is_connect() {
                    "connect"
                } else {
                    "transport"
                }
                .to_string(),
                message: error.to_string(),
                source_url: request.url.clone(),
            })?;
            let status = response.status().as_u16();
            let final_url = response.url().to_string();
            let headers = RECEIPT_HEADERS
                .iter()
                .filter_map(|name| {
                    response
                        .headers()
                        .get(*name)
                        .and_then(|value| value.to_str().ok())
                        .map(|value| ((*name).to_string(), value.to_string()))
                })
                .collect();
            let body = response
                .bytes()
                .await
                .map_err(|error| TransportError {
                    kind: "body".to_string(),
                    message: error.to_string(),
                    source_url: request.url,
                })?
                .to_vec();
            Ok(ResponseReceipt {
                body,
                final_url,
                status,
                headers,
                diagnostics: TransportDiagnostics::response(),
            })
        })
    }
}

/// Capture an HTTP response through caller-supplied transport without decoding it.
pub async fn capture_response_with_transport(
    request: TransportRequest,
    transport: &dyn Transport,
) -> std::result::Result<ResponseReceipt, TransportError> {
    transport.execute(request).await
}

/// Capture an HTTP response with the default reqwest transport.
#[cfg(feature = "runtime")]
pub async fn capture_response(
    request: TransportRequest,
) -> std::result::Result<ResponseReceipt, TransportError> {
    capture_response_with_transport(request, &ReqwestTransport::default()).await
}
